lib/ecco/save_event_listener.rb

```diff
@@ -2,32 +2,31 @@
 
 module Ecco
   class SaveEventListener < EventListener
-    SAVE_EVENTS = {
-      EventType::QUERY           => "QUERY",
-      EventType::ROTATE          => "ROTATE",
-      EventType::WRITE_ROWS      => "WRITE_ROWS",
-      EventType::EXT_WRITE_ROWS  => "WRITE_ROWS",
-      EventType::UPDATE_ROWS     => "UPDATE_ROWS",
-      EventType::EXT_UPDATE_ROWS => "UPDATE_ROWS",
-      EventType::DELETE_ROWS     => "DELETE_ROWS",
-      EventType::EXT_DELETE_ROWS => "DELETE_ROWS",
-    }
-
-    def accepted_events
-      SAVE_EVENTS.keys
+    def initialize(client)
+      events_of_interest = retrieve_save_events
+      super(client, events_of_interest)
     end
 
     def on_event(event)
       type = event.get_header.get_event_type
-
       case type
       when *accepted_events
         filename        = @client.get_binlog_filename
         position        = @client.get_binlog_position
-        event_type_name = SAVE_EVENTS.fetch(type)
+        event_type_name = event_type_to_string(type)
 
         @callback.call(filename, position, event_type_name)
       end
     end
+
+    private
+
+    def retrieve_save_events
+      Hash.new.merge(WRITE_EVENTS)
+              .merge(UPDATE_EVENTS)
+              .merge(DELETE_EVENTS)
+              .merge(QUERY_EVENTS)
+              .merge(ROTATE_EVENTS)
```

Comment:
Nice, but I think we should use the same helper method in both event handlers to keep it consistent. Also one could do something like this to avoid the list of merges (they are till there, but hidden):

```diff
index e59ccc1..6d72a17 100644
--- a/lib/ecco/event_listener.rb
+++ b/lib/ecco/event_listener.rb
@@ -15,7 +15,7 @@ module Ecco
 
     def initialize(client, events_of_interest)
       @client = client
-      @events_of_interest = events_of_interest
+      @events_of_interest = events_of_interest.inject(&:merge)
       @callback = Proc.new {}
     end
 
diff --git a/lib/ecco/row_event_listener.rb b/lib/ecco/row_event_listener.rb
index 5dcc2fe..6991c3a 100644
--- a/lib/ecco/row_event_listener.rb
+++ b/lib/ecco/row_event_listener.rb
@@ -4,8 +4,6 @@ require "ecco/row_event"
 module Ecco
   class RowEventListener < EventListener
     def initialize(client)
-      events_of_interest =
-        WRITE_EVENTS.merge(UPDATE_EVENTS).merge(DELETE_EVENTS)
       super(client, events_of_interest)
     end
 
@@ -32,5 +30,11 @@ module Ecco
         @callback.call(row_event)
       end
     end
+
+    private
+
+    def events_of_interest
+      [WRITE_EVENTS, UPDATE_EVENTS, DELETE_EVENTS]
+    end
   end
 end
diff --git a/lib/ecco/save_event_listener.rb b/lib/ecco/save_event_listener.rb
index dca57a0..8281969 100644
--- a/lib/ecco/save_event_listener.rb
+++ b/lib/ecco/save_event_listener.rb
@@ -3,7 +3,6 @@ require "ecco/event_listener"
 module Ecco
   class SaveEventListener < EventListener
     def initialize(client)
-      events_of_interest = retrieve_save_events
       super(client, events_of_interest)
     end
 
@@ -21,12 +20,8 @@ module Ecco
 
     private
 
-    def retrieve_save_events
-      Hash.new.merge(WRITE_EVENTS)
-              .merge(UPDATE_EVENTS)
-              .merge(DELETE_EVENTS)
-              .merge(QUERY_EVENTS)
-              .merge(ROTATE_EVENTS)
+    def events_of_interest
+      [WRITE_EVENTS, UPDATE_EVENTS, DELETE_EVENTS, QUERY_EVENTS, ROTATE_EVENTS]
     end
   end
 end
```

Reply:
Wow that's so cool! TIL!!

Reply:
One could actually further simplify, by doing this:

```diff
diff --git a/lib/ecco/event_listener.rb b/lib/ecco/event_listener.rb
index e59ccc1..62ebd81 100644
--- a/lib/ecco/event_listener.rb
+++ b/lib/ecco/event_listener.rb
@@ -13,10 +13,10 @@ module Ecco
 
     attr_writer :callback
 
-    def initialize(client, events_of_interest)
-      @client = client
-      @events_of_interest = events_of_interest
-      @callback = Proc.new {}
+    def initialize(client)
+      @client     = client
+      @event_list = events_of_interest.inject(&:merge)
+      @callback   = Proc.new {}
     end
 
     def on_event(event)
@@ -24,11 +24,17 @@ module Ecco
     end
 
     def accepted_events
-      @events_of_interest.keys
+      @event_list.keys
     end
 
     def event_type_to_string(type)
-      @events_of_interest.fetch(type)
+      @event_list.fetch(type)
+    end
+
+    private
+
+    def events_of_interest
+      raise NotImplementedError
     end
   end
 end
diff --git a/lib/ecco/row_event_listener.rb b/lib/ecco/row_event_listener.rb
index 5dcc2fe..26731a8 100644
--- a/lib/ecco/row_event_listener.rb
+++ b/lib/ecco/row_event_listener.rb
@@ -3,12 +3,6 @@ require "ecco/row_event"
 
 module Ecco
   class RowEventListener < EventListener
-    def initialize(client)
-      events_of_interest =
-        WRITE_EVENTS.merge(UPDATE_EVENTS).merge(DELETE_EVENTS)
-      super(client, events_of_interest)
-    end
-
     def on_event(event)
       data = event.get_data
       type = event.get_header.get_event_type
@@ -32,5 +26,11 @@ module Ecco
         @callback.call(row_event)
       end
     end
+
+    private
+
+    def events_of_interest
+      [WRITE_EVENTS, UPDATE_EVENTS, DELETE_EVENTS]
+    end
   end
 end
diff --git a/lib/ecco/save_event_listener.rb b/lib/ecco/save_event_listener.rb
index dca57a0..912709d 100644
--- a/lib/ecco/save_event_listener.rb
+++ b/lib/ecco/save_event_listener.rb
@@ -2,11 +2,6 @@ require "ecco/event_listener"
 
 module Ecco
   class SaveEventListener < EventListener
-    def initialize(client)
-      events_of_interest = retrieve_save_events
-      super(client, events_of_interest)
-    end
-
     def on_event(event)
       type = event.get_header.get_event_type
       case type
@@ -21,12 +16,8 @@ module Ecco
 
     private
 
-    def retrieve_save_events
-      Hash.new.merge(WRITE_EVENTS)
-              .merge(UPDATE_EVENTS)
-              .merge(DELETE_EVENTS)
-              .merge(QUERY_EVENTS)
-              .merge(ROTATE_EVENTS)
+    def events_of_interest
+      [WRITE_EVENTS, UPDATE_EVENTS, DELETE_EVENTS, QUERY_EVENTS, ROTATE_EVENTS]
     end
   end
 end
```

Reply:
Also, I think we should take care of `#keys` not having to be evaluated every time, that could just be something like (in addition to diff above):


```diff
diff --git a/lib/ecco/event_listener.rb b/lib/ecco/event_listener.rb
index 62ebd81..bd9e5d9 100644
--- a/lib/ecco/event_listener.rb
+++ b/lib/ecco/event_listener.rb
@@ -14,9 +14,10 @@ module Ecco
     attr_writer :callback
 
     def initialize(client)
-      @client     = client
-      @event_list = events_of_interest.inject(&:merge)
-      @callback   = Proc.new {}
+      @client          = client
+      @event_list      = events_of_interest.inject(&:merge)
+      @accepted_events = @event_list.keys
+      @callback        = Proc.new {}
     end
 
     def on_event(event)
@@ -24,7 +25,7 @@ module Ecco
     end
 
     def accepted_events
-      @event_list.keys
+      @accepted_events
     end
 
     def event_type_to_string(type)
```

Reply:
Added all the ideas in the comments here: 3dba86e644920cc9c75b7458674f341393e9058c